### PR TITLE
Add quoted patterns

### PR DESCRIPTION
**Pull request by odersky**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -1930,17 +1930,55 @@ class Typer extends Namer
...
         ctx.warning("Canceled splice directly inside a quote. '{ ${ XYZ } } is equivalent to XYZ.", tree.sourcePos)
         typed(innerExpr, pt)
-      case t if t.isType =>
-        typedTypeApply(untpd.TypeApply(untpd.ref(defn.InternalQuoted_typeQuoteR), List(tree.t)), pt)(quoteContext).withSpan(tree.span)
-      case t=>
-        typedApply(untpd.Apply(untpd.ref(defn.InternalQuoted_exprQuoteR), tree.t), pt)(quoteContext).withSpan(tree.span)
+      case quoted if quoted.isType =>
+        typedTypeApply(untpd.TypeApply(untpd.ref(defn.InternalQuoted_typeQuoteR), quoted :: Nil), pt)(quoteContext).withSpan(tree.span)
+      case quoted =>
+        if (ctx.mode.is(Mode.Pattern)) {
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: We should only make this transformation at `level == 0`

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -1949,7 +1987,14 @@ class Typer extends Namer
         ctx.warning("Canceled quote directly inside a splice. ${ '{ XYZ } } is equivalent to XYZ.", tree.sourcePos)
         typed(innerExpr, pt)
       case expr =>
-        typedApply(untpd.Apply(untpd.ref(defn.InternalQuoted_exprSpliceR), tree.expr), pt)(spliceContext).withSpan(tree.span)
+        if (ctx.mode.is(Mode.QuotedPattern)) {
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: We should only do this transformation at `level == 1` (i.e. the contents of the splice are at level 0)

File: library/src-bootstrapped/scala/internal/Quoted.scala [hunk 3]
```diff
@@ -16,4 +16,7 @@ object Quoted {
   def typeQuote[T <: AnyKind]: Type[T] =
     throw new Error("Internal error: this method call should have been replaced by the compiler")
 
+  /** A splice in a quoted pattern is desugared by the compiler into a call to this method */
+  def patternHole[T]: T =
```
Comment (nicolasstucki) [inline library/src-bootstrapped/scala/internal/Quoted.scala]: Note: this is the equivalent to [`def hole[T]`](https://github.com/lampepfl/dotty/pull/6042/files#diff-c60b7bdb36081577552ac360877d74ebR22) in #6042, not to be confused with [`object patternHole { def unapply[T] = ... }`](https://github.com/lampepfl/dotty/pull/6042/files#diff-c60b7bdb36081577552ac360877d74ebR26).

**Comment by odersky** [comment]
Status: FromTasty test fails on `quotedPatterns.scala`, presumably because there are (unreachable) warnings. Otherwise we seem to be good.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Agreed.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Agreed

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 4]
```diff
@@ -1930,17 +1930,57 @@ class Typer extends Namer
...
+    }
+    val result = splitter.transform(quoted)
+    (result, splitter.patBuf.toList)
+  }
+
+  /** A hole the shape pattern of a quoted.Matcher.unapply, representing a splice */
+  def patternHole(splice: Tree)(implicit ctx: Context): Tree =
+    ref(defn.InternalQuoted_patternHoleR).appliedToType(splice.tpe).withSpan(splice.span)
+
+  def givenReflection(implicit ctx: Context): Tree = Literal(Constant(null)) // FIXME: fill in
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: We still need to find the given `scala.tasty.Reflection` in scope.

**Comment by odersky** [comment]
@nicolasstucki Your commits LGTM

**Pull request merged**